modules/auxiliary/scanner/scada/modbusclient.rb

```diff
@@ -370,6 +387,38 @@ def write_registers
     end
   end
 
+  def read_id
+    @function_code = 0x2b
+    objCnt = 0
+    objIdPos = 0
+    objLen = 0
+    print_status("Sending READ ID...")
+    response = send_frame(make_read_id_payload)
+    if response.nil?
+      print_error("No answer for READ ID")
+    elsif response.size < 9
+      print_error("response is not modbus conform")
+    elsif response.unpack("C*")[7] == (0x80 | @function_code)
+      handle_error(response)
+    elsif response.size < 16
+      print_error("response is too short for READ ID")
+    elsif response.unpack("C*")[7] == @function_code
+      objCnt = response[13].unpack("C")[0]
+      objIdPos=14
+      begin
+        objLen = response[objIdPos+1].unpack("C")[0]
+        value = response.slice(objIdPos+2,objLen).unpack("a*")[0]
+        objid=response[objIdPos].unpack("C")[0]
+        print_good("Object ID #{objid}: #{value}")
+        objCnt = objCnt-1
+        objIdPos = objIdPos + objLen + 2
```

Comment:
There is the possibility objIdPos could end up pointing outside the length of `response` here. I'd update the logic to additionally check that `objIdPos` is still less than the length of `response` and exit immediately if it ever gets bigger.

Reply:
Fixed in https://github.com/rapid7/metasploit-framework/pull/16020/commits/577f5f81b96eb2dc8fff083073547615c45105f6